### PR TITLE
new is_authorized signatures in FFI; support human schema format for request validation

### PR DESCRIPTION
**Pull request by cdisselkoen**
## Description of changes

- Starts applying the recommendations in #757, in particular, the new `is_authorized` signatures for FFI.  
- Enables the human schema format for request validation in FFI.  
- Adds tests for request validation in FFI (we didn't previously have any).
- Refactors `AuthorizationCall::get_components()` to do a better job collecting multiple errors before returning, rather than eagerly returning when the first error is encountered.

## Issue #, if available

#757 (but this PR implements only a small portion of #757's suggestions)

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

### CODE REVIEW

File: cedar-policy/src/frontend/is_authorized.rs [hunk 1]
```diff
@@ -324,88 +346,270 @@ fn constant_true() -> bool {
...
+        };
+        let (policies, entities) = match self.slice.try_into(schema.as_ref()) {
+            Ok((policies, entities)) => (Some(policies), Some(entities)),
+            Err(e) => {
+                errs.extend(e);
+                (None, None)
+            }
+        };
+        if errs.is_empty() {
+            // TODO: this ignores warnings
```
Comment (khieta) [inline cedar-policy/src/frontend/is_authorized.rs:491]: Do we want to adjust the APIs in #757 to allow returning non-fatal errors (i.e., warnings)?

File: cedar-policy/src/frontend/is_authorized.rs [hunk 2]
```diff
@@ -324,88 +346,270 @@ fn constant_true() -> bool {
...
+            let Some(q) = q else {
+                return Err(errs);
+            };
+            let Some(policies) = policies else {
+                return Err(errs);
+            };
+            let Some(entities) = entities else {
+                return Err(errs);
+            };
```
Comment (khieta) [inline cedar-policy/src/frontend/is_authorized.rs:500]: These cases shouldn't be possible with an empty `errs`, right?

**Review by khieta — COMMENTED** [review]
Thanks for getting the ball rolling on #757 😄

You'll likely need to update the Java too -- you might want to try that before merging this PR to make sure it won't be too painful.

File: cedar-policy/src/frontend/is_authorized.rs [hunk 3]
```diff
@@ -303,13 +326,12 @@ struct AuthorizationCall {
         tsify(optional, type = "Record<string, CedarValueJson>")
     )]
     context: HashMap<String, JsonValueWithNoDuplicateKeys>,
-    /// Optional schema in JSON format.
+    /// Optional schema.
     /// If present, this will inform the parsing: for instance, it will allow
     /// `__entity` and `__extn` escapes to be implicit, and it will error if
     /// attributes have the wrong types (e.g., string instead of integer).
-    #[serde(rename = "schema")]
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs:310]: Correct. It was renaming from `schema` to `schema`

File: cedar-policy/src/frontend/is_authorized.rs (hunk 2 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs:500]: Correct.  But I thought returning `Err([])` in this impossible case was better than panicking

File: cedar-policy/src/frontend/is_authorized.rs [hunk 4]
```diff
@@ -324,88 +346,270 @@ fn constant_true() -> bool {
...
+            .map(|p| parse_entity_uid(p, "principal"))
+            .transpose()
+        {
+            Ok(None) => None,
+            Ok(Some(euid)) => Some(euid),
+            Err(e) => {
+                errs.push(e);
+                None
+            }
+        };
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs]: Factored some of the logic out; let me know what you think

File: cedar-policy/src/frontend/is_authorized.rs (hunk 1 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs:484]: We probably should

**Comment by cdisselkoen** [comment]
I'm concerned that the downstream Java CI is not failing.  It should be, I think?

**Comment by cdisselkoen** [comment]
I wonder if the downstream Java CI is somehow not pulling the correct version of `cedar-policy` when it runs its tests.  I.e., not pulling in the version of `cedar-policy` from this PR's branch.  I can't immediately tell from the CI log what version/hash of `cedar-policy` was used.

File: cedar-policy/src/frontend/is_authorized.rs (hunk 1 above)
Comment (khieta) [reply cedar-policy/src/frontend/is_authorized.rs:485]: How about the following?
```rust
enum AuthorizationAnswer {
    Failure { errors: HashSet<serde_json::Value> },
    Success { 
        response: Response,
        warnings: HashSet<serde_json::Value>,
    },
}
```
The miette json format should support tagging an "error" with the severity "warning".

**Review by khieta — APPROVED** [review]

File: cedar-policy/src/frontend/is_authorized.rs (hunk 2 above)
Comment (khieta) [reply cedar-policy/src/frontend/is_authorized.rs]: It still feels a little ugly to me 🙂 would prefer if you could refactor to remove, but it's not a blocking issue.

File: cedar-policy/src/frontend/is_authorized.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (khieta) [reply cedar-policy/src/frontend/is_authorized.rs]: New structure looks better

File: cedar-policy/src/frontend/is_authorized.rs (hunk 1 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs:515]: Or, do we envision that `Response` (i.e., `Diagnostics`) will eventually evolve to include warnings?  Would it make more sense to put all warnings inside `Diagnostics`?

File: cedar-policy/src/frontend/is_authorized.rs (hunk 2 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs]: Refactored

File: cedar-policy/src/frontend/is_authorized.rs [hunk 5]
```diff
@@ -95,42 +91,42 @@ fn is_authorized_partial(call: AuthorizationCall) -> PartialAuthorizationAnswer
...
-        |call| match is_authorized_partial(call) {
-            answer @ PartialAuthorizationAnswer::Residuals { .. } => {
-                InterfaceResult::succeed(answer)
-            }
-            PartialAuthorizationAnswer::ParseFailed { errors } => {
-                InterfaceResult::fail_bad_request(errors)
-            }
-        },
-    )
+pub fn is_authorized_json_partial(
```
Comment (khieta) [inline cedar-policy/src/frontend/is_authorized.rs:98]: Nit
```suggestion
pub fn is_authorized_partial_json(
```

File: cedar-policy/src/frontend/is_authorized.rs [hunk 6]
```diff
@@ -95,42 +91,42 @@ fn is_authorized_partial(call: AuthorizationCall) -> PartialAuthorizationAnswer
...
+    let ans = is_authorized_partial(serde_json::from_value(json)?);
+    serde_json::to_value(ans)
 }
 
-/// Interface version of a `Response` that uses `InterfaceDiagnostics` for simpler (de)serialization
+/// Input and output are strings containing serialized JSON, in the shapes
+/// expected by [`is_authorized_json_partial()`]
+#[doc = include_str!("../../experimental_warning.md")]
+#[cfg(feature = "partial-eval")]
+pub fn is_authorized_json_str_partial(json: &str) -> Result<String, serde_json::Error> {
```
Comment (khieta) [inline cedar-policy/src/frontend/is_authorized.rs:109]:
```suggestion
pub fn is_authorized_partial_json_str(json: &str) -> Result<String, serde_json::Error> {
```

File: cedar-policy/src/frontend/is_authorized.rs [hunk 7]
```diff
@@ -305,27 +303,35 @@ impl TryFrom<PartialResponse> for InterfaceResidualResponse {
 #[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
 #[cfg_attr(feature = "wasm", tsify(into_wasm_abi, from_wasm_abi))]
 pub enum AuthorizationAnswer {
-    /// Represents a failure to parse and call the authorizer
-    Failure {
-        /// Parsing errors
+    /// Represents a failure to parse or call the authorizer entirely
+    ParseFailed {
```
Comment (khieta) [inline cedar-policy/src/frontend/is_authorized.rs:307]: I prefer the more general name `Failure`. I'm imagining that we'll want to use these types for a general DRT interface, and other implementations of Cedar (e.g., the Lean) may return errors for more than just JSON parsing issues.

**Review by khieta — APPROVED** [review]

File: cedar-policy/src/frontend/is_authorized.rs (hunk 7 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs:307]: That's fair.  I changed the name to `ParseFailed` as part of the merge from `main` (a recent commit to `main` changed the name to `ParseFailed` -- see https://github.com/cedar-policy/cedar/blob/38754319c7cd3b50083f44c4098cfb10f27d540c/cedar-policy/src/frontend/is_authorized.rs#L324), but I am happy to change back to `Failure` for the reasons you give

File: cedar-policy/src/frontend/is_authorized.rs (hunk 1 above)
Comment (khieta) [reply cedar-policy/src/frontend/is_authorized.rs]: Good question - I think that would also be reasonable.

File: cedar-policy/src/frontend/is_authorized.rs (hunk 7 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs]: Changed

**Comment by cdisselkoen** [comment]
The Java (and DRT) downstream failures are indeed the result of this PR.  I will prep fixes to those before merging this.

File: cedar-policy/src/frontend/is_authorized.rs (hunk 1 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs]: Tried the version with warnings in `Diagnostics` and updated this PR; let me know what you think.

Also accounted for the possibility that we have both errors and warnings, by adding a `warnings` field to `Failure` here as well.

Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs]: Having second thoughts.  In the fullness of time we probably want to support warnings generated by evaluating a particular policy, and also warnings generated when parsing/processing things like schemas and entity data.  The former are associated with a particular policy id and make sense to be in `Diagnostics`.  The latter are not and do not.  Perhaps we want a space for warnings both in `Diagnostics` (for warnings associated with evaluating a particular policy) and in `Failure` and `Success` generally.

Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs]: Adjusted again in 684e8f65 on this PR

File: cedar-policy/src/frontend/is_authorized.rs [hunk 8]
```diff
@@ -656,9 +825,47 @@ impl RecvdSlice {
 #[cfg(test)]
 mod test {
     use super::*;
-    use crate::{frontend::utils::assert_is_failure, EntityUid};
     use cool_asserts::assert_matches;
-    use std::collections::HashMap;
+    use serde_json::json;
+
+    /// Assert that `is_authorized_json()` returns Allow with no errors
+    #[track_caller] // report the caller's location as the location of the panic, not the location in this function
```
Comment (andrewmwells-amazon) [inline cedar-policy/src/frontend/is_authorized.rs:832]: Is it considered a best practice to have this comment every time? It feels redundant to me.

File: cedar-policy/src/frontend/is_authorized.rs [hunk 9]
```diff
@@ -370,51 +398,121 @@ fn constant_true() -> bool {
...
-fn parse_schema(
-    schema_json: Option<JsonValueWithNoDuplicateKeys>,
-) -> Result<Option<Schema>, Vec<String>> {
-    schema_json
-        .map(|v| Schema::from_json_value(v.into()))
-        .transpose()
-        .map_err(|e| vec![e.to_string()])
+#[derive(Debug, Serialize, Deserialize)]
+enum Schema {
+    /// Schema in Cedar's default, human-readable schema format. See <https://docs.cedarpolicy.com/schema/human-readable-schema.html>
```
Comment (andrewmwells-amazon) [inline cedar-policy/src/frontend/is_authorized.rs:403]: I thought we wanted to just call this `Cedar schema format` going forwards?

**Review by andrewmwells-amazon — APPROVED** [review]

File: cedar-policy/src/frontend/is_authorized.rs (hunk 8 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs:832]: Can remove :)

File: cedar-policy/src/frontend/is_authorized.rs (hunk 9 above)
Comment (cdisselkoen) [reply cedar-policy/src/frontend/is_authorized.rs:403]: Will change.

**Review by andrewmwells-amazon — APPROVED** [review]

**Pull request merged**